src/pages/puzzles/[instanceId]/index.tsx

```diff
@@ -38,35 +41,55 @@ const PuzzlePage = ({
     user
   ) => {
     if (isAuthenticated) {
-      const submission = await submitPuzzleInstance(
-        answer,
-        puzzleInstance,
-        puzzle,
-        randomSeed,
-        user
-      );
+      const submissionInstance = async () => {
+        const submission = await submitPuzzleInstance(
+          answer,
+          puzzleInstance,
+          puzzle,
+          randomSeed,
+          user
+        );
 
-      if ((submission as HandledError).error) {
-        // TODO: handle errors from submitting
-        alert(submission.message);
-        return;
-      }
+        if ((submission as HandledError).error) {
+          console.error((submission as HandledError).message);
+          throw new Error('Something went wrong :(');
+        } else {
+          const success = submission.isCorrect[submission.isCorrect.length - 1];
 
-      const success = submission.isCorrect.at(-1);
+          const feedbackGifs = success
+            ? allFeedbackGifs.correct
+            : allFeedbackGifs.incorrect;
 
-      const correctnessMessage = success
-        ? 'Correct! Nice work!'
-        : 'Almost there! Try Again.';
-      alert(`${correctnessMessage}`);
+          setIsRecentCorrect(success);
+          setFeedbackGifSrc(getRandomGifSrc(feedbackGifs, success));
 
-      const feedbackGifs = success
-        ? allFeedbackGifs.correct
-        : allFeedbackGifs.incorrect;
+          if (!success) {
+            throw new Error('Almost there! Try Again');
+          }
+        }
+      };
 
-      setIsRecentCorrect(success);
-      setFeedbackGifSrc(getRandomGifSrc(feedbackGifs, success));
+      toast.promise(submissionInstance(), {
+        loading: 'Submitting your answer... ⏳',
+        success: 'Correct! Nice work!',
+        error: err => {
+          return err.message;
+        }
+      });
     } else {
-      alert('You must be logged in to submit puzzles!');
+      toast.error(t => (
+        <span>
```

Comment:
Yeah idk how I feel about having a bunch of custom HTML, can't we make components to put inside this? And just have a few different types of toast components

Reply:
i'll do that!